swaynag.1.scd
swaynag(1)

# NAME

swaynag - Show a warning or error message with buttons

# SYNOPSIS

_swaynag_ [options...]

# OPTIONS

*-b, --button* <text> <action>
	Create a button with the text _text_ that executes _action_ when pressed.
	If the environment variable _TERMINAL_ is set, _action_ will be run inside
	the terminal. Otherwise, it will fallback to running directly. Multiple
	buttons can be defined by providing the flag multiple times.

*-B, --button-no-terminal* <text> <action>
	Create a button with the text _text_ that executes _action_ when pressed.
	_action_ will be run directly instead of in a terminal. Multiple buttons
	can be defined by providing the flag multiple times.

*-z, --button-dismiss* <text> <action>
	Create a button with the text _text_ that executes _action_ when pressed,
	and dismisses swaynag. If the environment variable _TERMINAL_ is set,
	_action_ will be run inside the terminal. Otherwise, it will fallback to
	running directly. Multiple buttons can be defined by providing the flag
	multiple times.

*-Z, --button-dismiss-no-terminal* <text> <action>
	Create a button with the text _text_ that executes _action_ when pressed,
	and dismisses swaynag. _action_ will be run directly instead of in a
	terminal. Multiple buttons can be defined by providing the flag multiple
	times.

*-d, --debug*
	Enable debugging.

*-e, --edge* top|bottom
	Set the edge to use.

*-y, --layer* overlay|top|bottom|background
	Set the layer to use.

*-f, --font* <font>
	Set the font to use.

*-h, --help*
	Show help message and quit.

*-l, --detailed-message*
	Read a detailed message from stdin. A button to toggle details will be
	added. Details are shown in a scrollable multi-line text area.

*-L, --detailed-button* <text>
	Set the text for the button that toggles details. This has no effect if
	there is not a detailed message. The default is _Toggle details_.

*-m, --message* <msg>
	Set the message text.

*-o, --output* <output>
	Set the output to use. This should be the name of a _xdg\_output_.

*-s, --dismiss-button* <text>
	Sets the text for the dismiss nagbar button. The default is _X_.

*-v, --version*
	Show the version number and quit.

# APPEARANCE OPTIONS

*--background* <RRGGBB[AA]>
	Set the color of the background.

*--border* <RRGGBB[AA]>
	Set the color of the border.

*--border-bottom* <RRGGBB[AA]>
	Set the color of the bottom border.

*--button-background* <RRGGBB[AA]>
	Set the color for the background for buttons.

*--text* <RRGGBB[AA]>
	Set the text color.

*--button-text* <RRGGBB[AA]>
	Set the button text color.

*--border-bottom-size* <size>
	Set the thickness of the bottom border.

*--message-padding* <padding>
	Set the padding for the message.

*--details-background* <RRGGBB[AA]>
	Set the color for the background for details.

*--details-border-size* <size>
	Set the thickness for the details border.

*--button-border-size* <size>
	Set the thickness for the button border.

*--button-gap* <gap>
	Set the size of the gap between buttons.

*--button-dismiss-gap* <gap>
	Set the size of the gap between the dismiss button and another button.

*--button-margin-right* <margin>
	Set the margin from the right of the dismiss button to edge.

*--button-padding* <padding>
	Set the padding for the button text.



swaynag.c
#include <assert.h>
#include <poll.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <sys/timerfd.h>
#include <sys/wait.h>
#include <unistd.h>
#include <wayland-client.h>
#include <wayland-cursor.h>
#include <wlr/util/log.h>
#include "render.h"
#include "swaynag.h"
#include "wlr-layer-shell-unstable-v1-client-protocol.h"

static void nop() {
	// Intentionally left blank
}

static bool terminal_execute(char *terminal, char *command) {
	char fname[] = "/tmp/swaynagXXXXXX";
	FILE *tmp= fdopen(mkstemp(fname), "w");
	if (!tmp) {
		wlr_log(WLR_ERROR, "Failed to create temp script");
		return false;
	}
	wlr_log(WLR_DEBUG, "Created temp script: %s", fname);
	fprintf(tmp, "#!/bin/sh\nrm %s\n%s", fname, command);
	fclose(tmp);
	chmod(fname, S_IRUSR | S_IWUSR | S_IXUSR);
	size_t cmd_size = strlen(terminal) + strlen(" -e ") + strlen(fname) + 1;
	char *cmd = malloc(cmd_size);
	if (!cmd) {
		perror("malloc");
		return false;
	}
	snprintf(cmd, cmd_size, "%s -e %s", terminal, fname);
	execlp("sh", "sh", "-c", cmd, NULL);
	wlr_log_errno(WLR_ERROR, "Failed to run command, execlp() returned.");
	free(cmd);
	return false;
}

static void swaynag_button_execute(struct swaynag *swaynag,
		struct swaynag_button *button) {
	wlr_log(WLR_DEBUG, "Executing [%s]: %s", button->text, button->action);
	if (button->type == SWAYNAG_ACTION_DISMISS) {
		swaynag->run_display = false;
	} else if (button->type == SWAYNAG_ACTION_EXPAND) {
		swaynag->details.visible = !swaynag->details.visible;
		render_frame(swaynag);
	} else {
		pid_t pid = fork();
		if (pid < 0) {
			wlr_log_errno(WLR_DEBUG, "Failed to fork");
			return;
		} else if (pid == 0) {
			// Child process. Will be used to prevent zombie processes
			pid = fork();
			if (pid < 0) {
				wlr_log_errno(WLR_DEBUG, "Failed to fork");
				return;
			} else if (pid == 0) {
				// Child of the child. Will be reparented to the init process
				char *terminal = getenv("TERMINAL");
				if (button->terminal && terminal && *terminal) {
					wlr_log(WLR_DEBUG, "Found $TERMINAL: %s", terminal);
					if (!terminal_execute(terminal, button->action)) {
						swaynag_destroy(swaynag);
						_exit(EXIT_FAILURE);
					}
				} else {
					if (button->terminal) {
						wlr_log(WLR_DEBUG,
								"$TERMINAL not found. Running directly");
					}
					execlp("sh", "sh", "-c", button->action, NULL);
					wlr_log_errno(WLR_DEBUG, "execlp failed");
					_exit(EXIT_FAILURE);
				}
			}
			_exit(EXIT_SUCCESS);
		}

		if (button->dismiss) {
			swaynag->run_display = false;
		}

		if (waitpid(pid, NULL, 0) < 0) {
			wlr_log_errno(WLR_DEBUG, "waitpid failed");
		}
	}
}

static void layer_surface_configure(void *data,
		struct zwlr_layer_surface_v1 *surface,
		uint32_t serial, uint32_t width, uint32_t height) {
	struct swaynag *swaynag = data;
	swaynag->width = width;
	swaynag->height = height;
	zwlr_layer_surface_v1_ack_configure(surface, serial);
	render_frame(swaynag);
}

static void layer_surface_closed(void *data,
		struct zwlr_layer_surface_v1 *surface) {
	struct swaynag *swaynag = data;
	swaynag_destroy(swaynag);
}

static const struct zwlr_layer_surface_v1_listener layer_surface_listener = {
	.configure = layer_surface_configure,
	.closed = layer_surface_closed,
};

static void surface_enter(void *data, struct wl_surface *surface,
		struct wl_output *output) {
	struct swaynag *swaynag = data;
	struct swaynag_output *swaynag_output;
	wl_list_for_each(swaynag_output, &swaynag->outputs, link) {
		if (swaynag_output->wl_output == output) {
			wlr_log(WLR_DEBUG, "Surface enter on output %s",
					swaynag_output->name);
			swaynag->output = swaynag_output;
			swaynag->scale = swaynag->output->scale;
			render_frame(swaynag);
			break;
		}
	};
}

static const struct wl_surface_listener surface_listener = {
	.enter = surface_enter,
	.leave = nop,
};

static void update_cursor(struct swaynag_seat *seat) {
	struct swaynag_pointer *pointer = &seat->pointer;
	struct swaynag *swaynag = seat->swaynag;
	if (pointer->cursor_theme) {
		wl_cursor_theme_destroy(pointer->cursor_theme);
	}
	const char *cursor_theme = getenv("XCURSOR_THEME");
	unsigned cursor_size = 24;
	const char *env_cursor_size = getenv("XCURSOR_SIZE");
	if (env_cursor_size && *env_cursor_size) {
		errno = 0;
		char *end;
		unsigned size = strtoul(env_cursor_size, &end, 10);
		if (!*end && errno == 0) {
			cursor_size = size;
		}
	}
	pointer->cursor_theme = wl_cursor_theme_load(
		cursor_theme, cursor_size * swaynag->scale, swaynag->shm);
	if (!pointer->cursor_theme) {
		wlr_log(WLR_ERROR, "Failed to load cursor theme");
		return;
	}
	struct wl_cursor *cursor = wl_cursor_theme_get_cursor(pointer->cursor_theme, "default");
	if (!cursor) {
		wlr_log(WLR_ERROR, "Failed to get default cursor from theme");
		return;
	}
	pointer->cursor_image = cursor->images[0];
	wl_surface_set_buffer_scale(pointer->cursor_surface,
			swaynag->scale);
	wl_surface_attach(pointer->cursor_surface,
			wl_cursor_image_get_buffer(pointer->cursor_image), 0, 0);
	wl_pointer_set_cursor(pointer->pointer, pointer->serial,
			pointer->cursor_surface,
			pointer->cursor_image->hotspot_x / swaynag->scale,
			pointer->cursor_image->hotspot_y / swaynag->scale);
	wl_surface_damage_buffer(pointer->cursor_surface, 0, 0,
			INT32_MAX, INT32_MAX);
	wl_surface_commit(pointer->cursor_surface);
}

void update_all_cursors(struct swaynag *swaynag) {
	struct swaynag_seat *seat;
	wl_list_for_each(seat, &swaynag->seats, link) {
		if (seat->pointer.pointer) {
			update_cursor(seat);
		}
	}
}

static void wl_pointer_enter(void *data, struct wl_pointer *wl_pointer,
		uint32_t serial, struct wl_surface *surface,
		wl_fixed_t surface_x, wl_fixed_t surface_y) {
	struct swaynag_seat *seat = data;

	struct swaynag_pointer *pointer = &seat->pointer;
	pointer->x = wl_fixed_to_int(surface_x);
	pointer->y = wl_fixed_to_int(surface_y);

	if (seat->swaynag->cursor_shape_manager) {
		struct wp_cursor_shape_device_v1 *device =
			wp_cursor_shape_manager_v1_get_pointer(
				seat->swaynag->cursor_shape_manager, wl_pointer);
		wp_cursor_shape_device_v1_set_shape(device, serial,
			WP_CURSOR_SHAPE_DEVICE_V1_SHAPE_DEFAULT);
		wp_cursor_shape_device_v1_destroy(device);
	} else {
		pointer->serial = serial;
		update_cursor(seat);
	}
}

static void wl_pointer_motion(void *data, struct wl_pointer *wl_pointer,
		uint32_t time, wl_fixed_t surface_x, wl_fixed_t surface_y) {
	struct swaynag_seat *seat = data;
	seat->pointer.x = wl_fixed_to_int(surface_x);
	seat->pointer.y = wl_fixed_to_int(surface_y);
}

static void wl_pointer_button(void *data, struct wl_pointer *wl_pointer,
		uint32_t serial, uint32_t time, uint32_t button, uint32_t state) {
	struct swaynag_seat *seat = data;
	struct swaynag *swaynag = seat->swaynag;

	if (state != WL_POINTER_BUTTON_STATE_PRESSED) {
		return;
	}

	double x = seat->pointer.x;
	double y = seat->pointer.y;

	struct swaynag_button *nagbutton;
	wl_list_for_each(nagbutton, &swaynag->buttons, link) {
		if (x >= nagbutton->x
				&& y >= nagbutton->y
				&& x < nagbutton->x + nagbutton->width
				&& y < nagbutton->y + nagbutton->height) {
			swaynag_button_execute(swaynag, nagbutton);
			return;
		}
	}

	if (swaynag->details.visible &&
			swaynag->details.total_lines != swaynag->details.visible_lines) {
		struct swaynag_button button_up = swaynag->details.button_up;
		if (x >= button_up.x
				&& y >= button_up.y
				&& x < button_up.x + button_up.width
				&& y < button_up.y + button_up.height
				&& swaynag->details.offset > 0) {
			swaynag->details.offset--;
			render_frame(swaynag);
			return;
		}

		struct swaynag_button button_down = swaynag->details.button_down;
		int bot = swaynag->details.total_lines;
		bot -= swaynag->details.visible_lines;
		if (x >= button_down.x
				&& y >= button_down.y
				&& x < button_down.x + button_down.width
				&& y < button_down.y + button_down.height
				&& swaynag->details.offset < bot) {
			swaynag->details.offset++;
			render_frame(swaynag);
			return;
		}
	}
}

static void wl_pointer_axis(void *data, struct wl_pointer *wl_pointer,
		uint32_t time, uint32_t axis, wl_fixed_t value) {
	struct swaynag_seat *seat = data;
	struct swaynag *swaynag = seat->swaynag;
	if (!swaynag->details.visible
			|| seat->pointer.x < swaynag->details.x
			|| seat->pointer.y < swaynag->details.y
			|| seat->pointer.x >= swaynag->details.x + swaynag->details.width
			|| seat->pointer.y >= swaynag->details.y + swaynag->details.height
			|| swaynag->details.total_lines == swaynag->details.visible_lines) {
		return;
	}

	int direction = wl_fixed_to_int(value);
	int bot = swaynag->details.total_lines - swaynag->details.visible_lines;
	if (direction < 0 && swaynag->details.offset > 0) {
		swaynag->details.offset--;
	} else if (direction > 0 && swaynag->details.offset < bot) {
		swaynag->details.offset++;
	}

	render_frame(swaynag);
}

static const struct wl_pointer_listener pointer_listener = {
	.enter = wl_pointer_enter,
	.leave = nop,
	.motion = wl_pointer_motion,
	.button = wl_pointer_button,
	.axis = wl_pointer_axis,
	.frame = nop,
	.axis_source = nop,
	.axis_stop = nop,
	.axis_discrete = nop,
};

static void seat_handle_capabilities(void *data, struct wl_seat *wl_seat,
		enum wl_seat_capability caps) {
	struct swaynag_seat *seat = data;
	bool cap_pointer = caps & WL_SEAT_CAPABILITY_POINTER;
	if (cap_pointer && !seat->pointer.pointer) {
		seat->pointer.pointer = wl_seat_get_pointer(wl_seat);
		wl_pointer_add_listener(seat->pointer.pointer,
				&pointer_listener, seat);
	} else if (!cap_pointer && seat->pointer.pointer) {
		wl_pointer_destroy(seat->pointer.pointer);
		seat->pointer.pointer = NULL;
	}
}

static const struct wl_seat_listener seat_listener = {
	.capabilities = seat_handle_capabilities,
	.name = nop,
};

static void output_scale(void *data, struct wl_output *output,
		int32_t factor) {
	struct swaynag_output *swaynag_output = data;
	swaynag_output->scale = factor;
	if (swaynag_output->swaynag->output == swaynag_output) {
		swaynag_output->swaynag->scale = swaynag_output->scale;
		if (!swaynag_output->swaynag->cursor_shape_manager) {
			update_all_cursors(swaynag_output->swaynag);
		}
		render_frame(swaynag_output->swaynag);
	}
}

static void output_name(void *data, struct wl_output *output,
		const char *name) {
	struct swaynag_output *swaynag_output = data;
	swaynag_output->name = strdup(name);

	const char *outname = swaynag_output->swaynag->conf->output;
	if (!swaynag_output->swaynag->output && outname &&
			strcmp(outname, name) == 0) {
		wlr_log(WLR_DEBUG, "Using output %s", name);
		swaynag_output->swaynag->output = swaynag_output;
	}
}

static const struct wl_output_listener output_listener = {
	.geometry = nop,
	.mode = nop,
	.done = nop,
	.scale = output_scale,
	.name = output_name,
	.description = nop,
};

static void handle_global(void *data, struct wl_registry *registry,
		uint32_t name, const char *interface, uint32_t version) {
	struct swaynag *swaynag = data;
	if (strcmp(interface, wl_compositor_interface.name) == 0) {
		swaynag->compositor = wl_registry_bind(registry, name,
				&wl_compositor_interface, 4);
	} else if (strcmp(interface, wl_seat_interface.name) == 0) {
		struct swaynag_seat *seat =
			calloc(1, sizeof(struct swaynag_seat));
		if (!seat) {
			perror("calloc");
			return;
		}

		seat->swaynag = swaynag;
		seat->wl_name = name;
		seat->wl_seat =
			wl_registry_bind(registry, name, &wl_seat_interface, 1);

		wl_seat_add_listener(seat->wl_seat, &seat_listener, seat);

		wl_list_insert(&swaynag->seats, &seat->link);
	} else if (strcmp(interface, wl_shm_interface.name) == 0) {
		swaynag->shm = wl_registry_bind(registry, name, &wl_shm_interface, 1);
	} else if (strcmp(interface, wl_output_interface.name) == 0) {
		if (!swaynag->output) {
			struct swaynag_output *output =
				calloc(1, sizeof(struct swaynag_output));
			if (!output) {
				perror("calloc");
				return;
			}
			output->wl_output = wl_registry_bind(registry, name,
					&wl_output_interface, 4);
			output->wl_name = name;
			output->scale = 1;
			output->swaynag = swaynag;
			wl_list_insert(&swaynag->outputs, &output->link);
			wl_output_add_listener(output->wl_output,
					&output_listener, output);
		}
	} else if (strcmp(interface, zwlr_layer_shell_v1_interface.name) == 0) {
		swaynag->layer_shell = wl_registry_bind(
				registry, name, &zwlr_layer_shell_v1_interface, 1);
	} else if (strcmp(interface, wp_cursor_shape_manager_v1_interface.name) == 0) {
		swaynag->cursor_shape_manager = wl_registry_bind(
				registry, name, &wp_cursor_shape_manager_v1_interface, 1);
	}
}

void swaynag_seat_destroy(struct swaynag_seat *seat) {
	if (seat->pointer.cursor_theme) {
		wl_cursor_theme_destroy(seat->pointer.cursor_theme);
	}
	if (seat->pointer.pointer) {
		wl_pointer_destroy(seat->pointer.pointer);
	}
	wl_seat_destroy(seat->wl_seat);
	wl_list_remove(&seat->link);
	free(seat);
}

static void handle_global_remove(void *data, struct wl_registry *registry,
		uint32_t name) {
	struct swaynag *swaynag = data;
	if (swaynag->output->wl_name == name) {
		swaynag->run_display = false;
	}

	struct swaynag_seat *seat, *tmpseat;
	wl_list_for_each_safe(seat, tmpseat, &swaynag->seats, link) {
		if (seat->wl_name == name) {
			swaynag_seat_destroy(seat);
		}
	}
}

static const struct wl_registry_listener registry_listener = {
	.global = handle_global,
	.global_remove = handle_global_remove,
};

void swaynag_setup_cursors(struct swaynag *swaynag) {
	struct swaynag_seat *seat;

	wl_list_for_each(seat, &swaynag->seats, link) {
		struct swaynag_pointer *p = &seat->pointer;

		p->cursor_surface =
			wl_compositor_create_surface(swaynag->compositor);
		assert(p->cursor_surface);
	}
}

void swaynag_setup(struct swaynag *swaynag) {
	swaynag->display = wl_display_connect(NULL);
	if (!swaynag->display) {
		wlr_log(WLR_ERROR, "Unable to connect to the compositor. "
				"If your compositor is running, check or set the "
				"WAYLAND_DISPLAY environment variable.");
		exit(EXIT_FAILURE);
	}

	swaynag->scale = 1;

	struct wl_registry *registry = wl_display_get_registry(swaynag->display);
	wl_registry_add_listener(registry, &registry_listener, swaynag);
	if (wl_display_roundtrip(swaynag->display) < 0) {
		wlr_log(WLR_ERROR, "failed to register with the wayland display");
		exit(EXIT_FAILURE);
	}

	assert(swaynag->compositor && swaynag->layer_shell && swaynag->shm);

	// Second roundtrip to get wl_output properties
	if (wl_display_roundtrip(swaynag->display) < 0) {
		wlr_log(WLR_ERROR, "Error during outputs init.");
		swaynag_destroy(swaynag);
		exit(EXIT_FAILURE);
	}

	if (!swaynag->output && swaynag->conf->output) {
		wlr_log(WLR_ERROR, "Output '%s' not found", swaynag->conf->output);
		swaynag_destroy(swaynag);
		exit(EXIT_FAILURE);
	}

	if (!swaynag->cursor_shape_manager) {
		swaynag_setup_cursors(swaynag);
	}

	swaynag->surface = wl_compositor_create_surface(swaynag->compositor);
	assert(swaynag->surface);
	wl_surface_add_listener(swaynag->surface, &surface_listener, swaynag);

	swaynag->layer_surface = zwlr_layer_shell_v1_get_layer_surface(
			swaynag->layer_shell, swaynag->surface,
			swaynag->output ? swaynag->output->wl_output : NULL,
			swaynag->conf->layer,
			"swaynag");
	assert(swaynag->layer_surface);
	zwlr_layer_surface_v1_add_listener(swaynag->layer_surface,
			&layer_surface_listener, swaynag);
	zwlr_layer_surface_v1_set_anchor(swaynag->layer_surface,
			swaynag->conf->anchors);

	wl_registry_destroy(registry);
}

void swaynag_run(struct swaynag *swaynag) {
	swaynag->run_display = true;
	render_frame(swaynag);
	int timer_fd = -1;
	if (swaynag->details.close_timeout != 0) {
		timer_fd = timerfd_create(CLOCK_MONOTONIC, TFD_CLOEXEC);
		struct itimerspec timeout = {
			.it_value.tv_sec = swaynag->details.close_timeout,
		};
		timerfd_settime(timer_fd, 0, &timeout, NULL);
	}
	struct pollfd fds[] = {
		{
			.fd = wl_display_get_fd(swaynag->display),
			.events = POLLIN,
		}, {
			.fd = timer_fd,
			.events = POLLIN,
		}
	};
	while (swaynag->run_display) {
		while (wl_display_prepare_read(swaynag->display) != 0) {
			wl_display_dispatch_pending(swaynag->display);
		}
		poll(fds, 2, -1);
		if (fds[0].revents & POLLIN) {
			if (timer_fd >= 0 && swaynag->details.close_timeout_cancel) {
				close(timer_fd);
				timer_fd = -1;
				fds[1].fd = -1;
			}
			wl_display_read_events(swaynag->display);
		} else {
			wl_display_cancel_read(swaynag->display);
		}
		if (timer_fd >= 0 && fds[1].revents & POLLIN) {
			close(timer_fd);
			timer_fd = -1;
			fds[1].fd = -1;
			swaynag->run_display = false;
		}
	}
}

void swaynag_destroy(struct swaynag *swaynag) {
	swaynag->run_display = false;

	free(swaynag->message);

	struct swaynag_button *button, *next;
	wl_list_for_each_safe(button, next, &swaynag->buttons, link) {
		wl_list_remove(&button->link);
		free(button->text);
		free(button->action);
		free(button);
	}
	free(swaynag->details.message);
	free(swaynag->details.details_text);
	free(swaynag->details.button_up.text);
	free(swaynag->details.button_down.text);

	if (swaynag->layer_surface) {
		zwlr_layer_surface_v1_destroy(swaynag->layer_surface);
	}

	if (swaynag->surface) {
		wl_surface_destroy(swaynag->surface);
	}

	struct swaynag_seat *seat, *tmpseat;
	wl_list_for_each_safe(seat, tmpseat, &swaynag->seats, link) {
		swaynag_seat_destroy(seat);
	}

	destroy_buffer(&swaynag->buffers[0]);
	destroy_buffer(&swaynag->buffers[1]);

	if (swaynag->outputs.prev || swaynag->outputs.next) {
		struct swaynag_output *output, *temp;
		wl_list_for_each_safe(output, temp, &swaynag->outputs, link) {
			wl_output_destroy(output->wl_output);
			free(output->name);
			wl_list_remove(&output->link);
			free(output);
		};
	}

	if (swaynag->compositor) {
		wl_compositor_destroy(swaynag->compositor);
	}

	if (swaynag->shm) {
		wl_shm_destroy(swaynag->shm);
	}

	if (swaynag->display) {
		wl_display_disconnect(swaynag->display);
	}
}
